### ISSUE TITLE
Image Policy does not work as expected

### ISSUE THREAD

**Issue by Jooho** [openshift/origin#15709]
From this [doc](https://docs.openshift.com/container-platform/3.5/admin_guide/image_policy.html), I can get sample for image policy. 

After restarting cluster, I expected : 
 - it can pull any images from docker.io 
 - it block imageStreamTag that has specific annotation 
 - it still can use default template like cakephp or eap based on imageStream

However, it can not possible to deploy default template. In detail, I can not build/deploy cakephp(using default template) anymore because it is blocked by image policy.
I am not sure it is expected behaviour? 

**Issues** :
- If I add matchRegistries, default templates are blocked by image policy
- If I add matchRegistries, matchIntegratedRegistry is ignored

**Workaround:**
- Add following lines as a image policy
~~~
        - name: allow-images-from-dockerhub
          onResources:
          - resource: pods
          - resource: deploymentconfigs
          - resource: replicationcontrollers
          - resource: builds
          matchRegistries:
          - docker.io
          - ""
          - registry.access.redhat.com
          - 172.30.210.68:5000          # <===== integrated docker registry svc ip
~~~
##### Version
[provide output of the `openshift version` or `oc version` command]
oc v3.5.5.26
kubernetes v1.5.2+43a9be4
features: Basic-Auth GSSAPI Kerberos SPNEGO

Server https://master10.example.com:8443
openshift v3.5.5.26
kubernetes v1.5.2+43a9be4

##### Steps To Reproduce
1. Add following into master-config.yaml
~~~
admissionConfig:
  pluginConfig:
    openshift.io/ImagePolicy:
      configuration:
        kind: ImagePolicyConfig
        apiVersion: v1
        resolveImages: AttemptRewrite
        executionRules:
        - name: allow-images-from-internal-registry
          # allows images from the internal registry and tries to resolve them
          onResources:
          - resource: pods
          - resource: builds
          reject: false
          matchIntegratedRegistry: true
        - name: allow-images-from-dockerhub
          onResources:
          - resource: pods
          - resource: builds
          matchRegistries:
          - docker.io

~~~
2. oc new-app cakephp-mysql-example
Then, you can see these errors on monitoring tab.
![1 after-adding-image-policy](https://user-images.githubusercontent.com/2375181/29133891-ebe21040-7d02-11e7-993c-a033cce6107e.png)

##### Current Result
deployment/build/replicationcontrollers/pods are blocked by image policy

##### Expected Result
The images from internal docker registry are allowed to deploy on openshift
The images from docker.io are allowed to deploy on openshift
The images(or imageStream) in default templates are allowed to deploy on openshift

**Comment by rezie** [comment]
The images that are trying to be deployed according to your image at the bottom (which are Openshift's component images) need to be prefixed with a registry (e.g. dockerhub's) that your ImagePolicy accepts, e.g. the `allow-images-from-dockerhub` rule. You can modify this via the `imageConfig.format` stanza in your master config.

**Comment by Jooho** [comment]
@rezie thanks for update.

I add registry.access.redhat.com in ```imageConfig.format``` in master-config.yaml then now it is successfully deploying openshift component images but still images from integrated docker registry could not deploy.
```
        executionRules:
        - name: allow-images-from-internal-registry
          # allows images from the internal registry and tries to resolve them
          onResources:
          - resource: pods
          - resource: builds
          reject: false
          matchIntegratedRegistry: true
```
This image policy is to allow the images from integrated docker registry, right?

**Comment by Jooho** [comment]
Moreover, if the way adding ```registry.access.redhat.com``` for ````mageConfig.format``` is right, I think doc should explain it because without it, ocp cluster will stop deploying any images due to image policy. what do you think?

**Comment by rezie** [comment]
I had confusion on my end trying to deploy images from the internal registry as well - see #15475. It looks like there's currently a problem if you are deploying applications via a DeploymentConfig; if you deploy an image through a Pod directly, then there's no problems with using the integrated registry.

I'll defer the prefixing question to someone else, since I'm not entirely sure if that workaround is the correct (preferred?) way to handle this.

**Comment by bparees** [comment]
@mfojtik can you provide some hints here?  is this a bug? a doc issue? user error?

@Jooho it sounds like maybe you just need to use multiple match policies, one to allow the integrated registry, and another to allow specific external registries?

**Comment by Jooho** [comment]
@bparees Yes, I thought I can make it with combining several policies but it was not.
```
1. The images from internal docker registry are allowed to deploy on openshift
2. The images from docker.io are allowed to deploy on openshift
3. The images(or imageStream) in default templates are allowed to deploy on openshift
```
1 and 2 are possible and already have [a doc](https://docs.openshift.com/container-platform/3.5/admin_guide/image_policy.html) to achieve it . However, the thing is if I add the configuration for 1 and 2, the images(registry.access.redhat.com) in default template are also block. 

Therefore, @rezie  recommend me to add prefix(registry.access.redhat.com) using `imageConfig.format` in master-config.yaml and add `registry.access.redhat.com` under `matchRegistries`.  However, still it is not possible to pull from integrated docker registry because of [#15475](https://github.com/openshift/origin/issues/15475).  

In conclusion, if client follow [the image policy doc sample](https://docs.openshift.com/container-platform/3.5/admin_guide/image_policy.html), they will encounter several issues. So we need to explain what right way is to use image Policy something like following:

 If you want to use imagePolicy :

**NOTE**: 3.6 has [issue](https://github.com/openshift/origin/issues/15475) to pull images from  integrated docker registry with imagePolicy, hence please use it from 3.6.1

1.  you need to add registry.access.redhat.com in `imageConfig.format` in master-config.yaml

2.  you need to add registry.access.redhat.com under `matchRegistries`

**Comment by bparees** [comment]
>  you need to add registry.access.redhat.com in imageConfig.format in master-config.yaml

I don't understand what this has to do with anything.. this determines which infrastructure images your cluster will use, but it's not part of how the policy works.  Whatever prefix you put here, does need to be allowed by your image policy, but that's as far as that goes.  You didn't have to explicitly change this value to make your policy work.

>  you need to add registry.access.redhat.com under matchRegistries

having to do this certainly makes sense, registry.access.redhat.com is just another external registry, you have to explicitly allow it if you're going to start locking down which registries you allow access to.

So i'm still not seeing what about this isn't working as expected?

**Comment by rezie** [comment]
> I don't understand what this has to do with anything.. this determines which infrastructure images your cluster will use, but it's not part of how the policy works. Whatever prefix you put here, does need to be allowed by your image policy, but that's as far as that goes. You didn't have to explicitly change this value to make your policy work.

At least in my experience, I hit the same error as @Jooho, where the infrastructure images would fail to be deployed unless a prefix was explicitly defined. The error message I got was identical to the screenshot in the first post.

The reason for modifying `format` in the master-config is to ensure that the prefix is set by default (as users aren't able to manually supply that prefix when creating deployments). Without this, the example as provided in the [documentation](https://docs.openshift.com/container-platform/3.5/admin_guide/image_policy.html#image-policy-configuring-the-image-policy-admission-plug-in) will prevent the infrastructure images from being deployed when either of the registry matchers are enforced.

**Comment by bparees** [comment]
Ok, I think i understand now, thanks @rezie.  It's a bit orthogonal to the imagepolicy but it's a worthy note/gotcha.

**Comment by bparees** [comment]
@Jooho @rezie i've created a doc update here: https://github.com/openshift/openshift-docs/pull/5078

can you take a look and see if it provides the additional information you think was lacking?

**Comment by rezie** [comment]
That looks good.

I don't know if this would be TMI for this scenario, but this is also configurable via the Advanced Installation process through the [oreg*](https://github.com/openshift/openshift-ansible/blob/master/inventory/byo/hosts.origin.example#L166) variable

**Comment by bparees** [comment]
@rezie thanks, i think i'll leave the advanced bit out.

**Comment by Jooho** [comment]
@rezie @bparees By the way, one thing worried to my mind is if somebody try image policy with 3.6, they could not pull images from integrated docker registry so they will eventually encounter unacceptable situation.

**Comment by bparees** [comment]
@Jooho sorry can you elaborate?  (i'm just picking up the imagestream components so i apologize for my lack of knowledge)

**Comment by Jooho** [comment]
@bparees 

Sure.

First of all, I would like to say that if I just follow the documentation, I encounter this error message :
```
$ oc new-app --image-stream=python --code https://github.com/openshift/django-ex.git

$ oc get event |grep Failed
Failed create | Error creating: Pod "" is invalid: spec.containers[0].image: Forbidden: this image is prohibited by policy
```

It is because DC use ImageStreamTag(172.x.x.x:5000/test/test@sha256:ce220..........b29432)

From the issue [#15475](https://github.com/openshift/origin/issues/15475), @mfojtik  explains like following:
```
this happen because we resolve the image for DC from the imagestream which results into the image name (which is represented by digest or 'ID').
```
This issue is fixed and included to 3.6.1

In order to avoid this issue at the moment, we need to add integrated docker registry service ip under `matchRegistries` like
```
         matchRegistries:
          - docker.io
          - registry.access.redhat.com
          - 172.30.210.68:5000     <====== ADD
```

**Comment by bparees** [comment]
@Jooho ah, I see, thank you.  That sounds like a better candidate for a knowledge base article or release note than the doc, since it's basically a bug.

**Comment by Jooho** [comment]
@bparees I will create KCS about it.

**Comment by bparees** [comment]
@Jooho thanks!
